### PR TITLE
[SLA] PIM-5215: All product fields are exported

### PR DESCRIPTION
**Pull request by damien-carcel**
Even empty fields are exported in CSV product export and quick export

| Q                 | A
| ----------------- | ---
| Specs             | Yes
| Behats            | ok
| Blue CI           | Running
| Changelog updated | Yes
| Review and 2 GTM  |

### CODE REVIEW

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 1]
```diff
@@ -60,6 +67,8 @@ public function __construct(
      */
     public function process($product)
     {
+        $this->productBuilder->addMissingProductValues($product);
```
Comment (juliensnz) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: you should test if the productBuilder is not null

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php [hunk 2]
```diff
@@ -71,6 +78,8 @@ public function initialize()
      */
     public function process($product)
     {
+        $this->productBuilder->addMissingProductValues($product);
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessor.php]: same here

File: CHANGELOG-1.4.md [hunk 3]
```diff
@@ -1,3 +1,8 @@
+# 1.4.* (2015-**-**)
```
Comment (jjanvier) [inline CHANGELOG-1.4.md]: change that, the tag script will fail

File: features/export/export_products.feature [hunk 4]
```diff
@@ -213,10 +213,11 @@ Feature: Export products
     When I am on the "footwear_product_export" export job page
     And I launch the export job
     And I wait for the "footwear_product_export" job to finish
+    And I wait 30 seconds
```
Comment (jjanvier) [inline features/export/export_products.feature]: no

Comment (willy-ahva) [inline features/export/export_products.feature]: oups

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php (hunk 1 above)
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: are we sure it does not degrade the performances drastically?

Comment (willy-ahva) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: @jjanvier seen with @nidup

File: spec/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php [hunk 5]
```diff
@@ -6,6 +6,7 @@
 use Akeneo\Bundle\BatchBundle\Entity\StepExecution;
 use Akeneo\Bundle\BatchBundle\Item\InvalidItemException;
 use PhpSpec\ObjectBehavior;
+use Pim\Bundle\CatalogBundle\Builder\ProductBuilder;
```
Comment (pierallard) [inline spec/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductToFlatArrayProcessorSpec.php]: ProductBuilderInterface?

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 6]
```diff
@@ -40,19 +41,25 @@ class ProductToFlatArrayProcessor extends AbstractConfigurableStepElement implem
...
+    protected $productBuilder;
+
     /**
-     * @param Serializer     $serializer
-     * @param ChannelManager $channelManager
-     * @param string[]       $mediaAttributeTypes
+     * @param Serializer       $serializer
+     * @param ChannelManager   $channelManager
+     * @param string[]         $mediaAttributeTypes
+     * @param ProductBuilderInterface $productBuilder
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: Alignment

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 7]
```diff
@@ -40,19 +41,25 @@ class ProductToFlatArrayProcessor extends AbstractConfigurableStepElement implem
...
      */
     public function __construct(
         Serializer $serializer,
         ChannelManager $channelManager,
-        array $mediaAttributeTypes
+        array $mediaAttributeTypes,
+        ProductBuilderInterface $productBuilder = null
```
Comment (pierallard) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: You have to be careful on the merge from 1.4 to master, we've added parameters to this constructor too.

File: spec/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessorSpec.php [hunk 8]
```diff
@@ -5,6 +5,7 @@
 use Akeneo\Component\FileStorage\Model\FileInfoInterface;
 use League\Flysystem\Filesystem;
 use PhpSpec\ObjectBehavior;
+use Pim\Bundle\CatalogBundle\Builder\ProductBuilderInterface;
```
Comment (nidup) [inline spec/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessorSpec.php]: take care when merge on 1.5 this one has been moved to Pim\Component\Catalog\Builder\ProductBuilderInterface;

File: src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php [hunk 9]
```diff
@@ -60,6 +67,10 @@ public function __construct(
      */
     public function process($product)
     {
+        if (null !== $this->productBuilder) {
+            $this->productBuilder->addMissingProductValues($product);
```
Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: I understand the purpose, avoid to create null values in database and add them dynamically during export.

However, for default product export, we always export values for a dedicated channel means here we can generate values for many channels to not export them at the end (selected channel and locales are into $this->getNormalizerContext() )

In other hand in the context of a patch not easy to re-work this.

Comment (nidup) [inline src/Pim/Bundle/BaseConnectorBundle/Processor/ProductToFlatArrayProcessor.php]: We need to only add global values (not localizable, not scopable) and localizable values (for configured locales of normalizercontext) and scopables values (for configured channel of normalizercontext).
GTM once this is done

**Comment by juliensnz** [comment]
GTM

**Comment by jjanvier** [comment]
gtm

**Pull request merged**